lib/location/method_cellular.c

```diff
@@ -136,11 +141,33 @@ static void method_cellular_positioning_work_fn(struct k_work *work)
 	/* NCELLMEAS done at this point of time. Store current time to response. */
 	location_utils_systime_to_location_datetime(&location_result.datetime);
 
+	/* Check if timeout is given */
+	if (cellular_config.timeout > 0) {
+		/* +1 to round the time up */
+		ncellmeas_time = ((k_uptime_get() - ncellmeas_start_time) / MSEC_PER_SEC) + 1;
+
+		/* Check if timeout has already elapsed */
+		if (ncellmeas_time >= cellular_config.timeout) {
+			LOG_WRN("Timeout occurred during neighbour cell measurement");
+			location_core_event_cb_timeout();
```

Comment:
As long as the system workqueue thread is at a non-preempt priority level, this should be true, yeah. But if cell measurement finishes at just the right time, it will be a flip of the coin whether this work function resumes (on line 129 at the semaphore), or whether the system workqueue initiates the timeout expiry function. I don't think that's a huge problem, but it'll create potentially inconsistent behavior


If someone changes the system workqueue thread priority to a preempting priority, then I'm pretty sure this routine could get preempted in the middle of manually detecting a timeout, and in that case all bets are off. For instance, the expiry function in location_core could preempt between lines 150 and 152, in which case you'll still get the double-call.

Reply:
Ah, yeah, I did not see that. You're right. This should work. I still find it a bit hard to follow, as currently written, and it makes me nervous that we have two separate timeout mechanisms on two separate threads, both of which are set up to execute at almost the exact same time. With non-preempting thread priorities it isn't a problem, but that is because the behavior deviates from intuition (while this work function is executing, the other timeout function literally cannot be executed, and vice-versa)

In fact, I am pretty sure that if cell measurement finishes at just the right time, it will be a coin-toss between whether this work function resumes on line 129, or the the location_core timeout work function initiates. That potential for inconsistency makes me nervous.

In general I think it would be cleaner to have only a single time-out mechanism rather than two timeout mechanisms that are running in parallel.

I guess the other option would be to remove the location_core timeout mechanism, and rely on the service integrations to implement the timeout on their own via their respective protocol layers. But I kind of feel like imposing the timeout at the location_core level is the cleaner approach.

Reply:
Yep, let's discuss in the other comment.